src/libraries/Microsoft.Extensions.DependencyInjection/src/ServiceLookup/ServiceProviderEngineScope.cs

```diff
@@ -155,30 +163,48 @@ static async ValueTask Await(int i, ValueTask vt, List<object> toDispose)
                         ((IDisposable)disposable).Dispose();
                     }
                 }
+
+                ClearState();
+            }
+        }
+
+        private IDictionary<ServiceCacheKey, object> ScopeDisposed()
+        {
+            ThrowHelper.ThrowObjectDisposedException();
+            return null;
+        }
+
+        private void ClearState()
+        {
+            // Dispose the state, which will end up attempting to return the state pool.
+            // This will return false if the pool is full or if this state object is the root scope
+            if (_state.Return())
```

Comment:
There's a race here I need to fix between getting resolved services and returning state to the pool.